Ignore files in .gitignore in tidy

jyn514 opened the pull request:
- Switch from `walkdir` to `ignore`. This required various changes to make `skip` thread-safe.
- Ignore `build` anywhere in the source tree, not just at the top-level. We support this in bootstrap, we should support it in tidy too.

As a nice side benefit, this also makes tidy a bit faster.

Before:
```
; hyperfine -i '"/home/gh-jyn514/rust2/build/aarch64-unknown-linux-gnu/stage0-tools-bin/rust-tidy" "/home/gh-jyn514/rust2" "/home/gh-jyn514/rust2/build/aarch64-unknown-linux-gnu/stage0/bin/cargo" "/home/gh-jyn514/rust2/build" "32"'
Benchmark 1: "/home/gh-jyn514/rust2/build/aarch64-unknown-linux-gnu/stage0-tools-bin/rust-tidy" "/home/gh-jyn514/rust2" "/home/gh-jyn514/rust2/build/aarch64-unknown-linux-gnu/stage0/bin/cargo" "/home/gh-jyn514/rust2/build" "32"
  Time (mean ± σ):      1.080 s ±  0.008 s    [User: 2.616 s, System: 3.243 s]
  Range (min … max):    1.069 s …  1.099 s    10 runs
```

After:
```
; hyperfine '"/home/gh-jyn514/rust2/build/aarch64-unknown-linux-gnu/stage0-tools-bin/rust-tidy" "/home/gh-jyn514/rust2" "/home/gh-jyn514/rust2/build/aarch64-unknown-linux-gnu/stage0/bin/cargo" "/home/gh-jyn514/rust2/build" "32"'
Benchmark 1: "/home/gh-jyn514/rust2/build/aarch64-unknown-linux-gnu/stage0-tools-bin/rust-tidy" "/home/gh-jyn514/rust2" "/home/gh-jyn514/rust2/build/aarch64-unknown-linux-gnu/stage0/bin/cargo" "/home/gh-jyn514/rust2/build" "32"
  Time (mean ± σ):     705.0 ms ±   1.4 ms    [User: 3179.1 ms, System: 1517.5 ms]
  Range (min … max):   702.3 ms … 706.9 ms    10 runs
```

r? @the8472

jyn514 committed:
Ignore things in .gitignore in tidy

- Switch from `walkdir` to `ignore`. This required various changes to
  make `skip` thread-safe.
- Ignore `build` anywhere in the source tree, not just at the top-level.
  We support this in bootstrap, we should support it in tidy too.

As a nice side benefit, this also makes tidy a bit faster.

Before:
```
; hyperfine -i '"/home/gh-jyn514/rust2/build/aarch64-unknown-linux-gnu/stage0-tools-bin/rust-tidy" "/home/gh-jyn514/rust2" "/home/gh-jyn514/rust2/build/aarch64-unknown-linux-gnu/stage0/bin/cargo" "/home/gh-jyn514/rust2/build" "32"'
Benchmark 1: "/home/gh-jyn514/rust2/build/aarch64-unknown-linux-gnu/stage0-tools-bin/rust-tidy" "/home/gh-jyn514/rust2" "/home/gh-jyn514/rust2/build/aarch64-unknown-linux-gnu/stage0/bin/cargo" "/home/gh-jyn514/rust2/build" "32"
  Time (mean ± σ):      1.080 s ±  0.008 s    [User: 2.616 s, System: 3.243 s]
  Range (min … max):    1.069 s …  1.099 s    10 runs
```

After:
```
; hyperfine '"/home/gh-jyn514/rust2/build/aarch64-unknown-linux-gnu/stage0-tools-bin/rust-tidy" "/home/gh-jyn514/rust2" "/home/gh-jyn514/rust2/build/aarch64-unknown-linux-gnu/stage0/bin/cargo" "/home/gh-jyn514/rust2/build" "32"'
Benchmark 1: "/home/gh-jyn514/rust2/build/aarch64-unknown-linux-gnu/stage0-tools-bin/rust-tidy" "/home/gh-jyn514/rust2" "/home/gh-jyn514/rust2/build/aarch64-unknown-linux-gnu/stage0/bin/cargo" "/home/gh-jyn514/rust2/build" "32"
  Time (mean ± σ):     705.0 ms ±   1.4 ms    [User: 3179.1 ms, System: 1517.5 ms]
  Range (min … max):   702.3 ms … 706.9 ms    10 runs
```

jyn514 commented:
I originally tried making all these checks parallel before realizing that you'd already done that at the top level 🤦 just not at the file level which is the thing I did. If you want the code it's here: https://github.com/rust-lang/rust/compare/master...jyn514:rust:parallel-tidy?expand=1)

but not sure how useful it is given that `walk` is within the checks and not at the top-level, seems pretty hard to be able to intersperse the checks with each other.

the8472 commented:
Switching to file level parallelism may be better on many-core systems because doing it at the top level is fairly coarse and the longest-running tasks (`style.rs` was the slowest last time I checked) end up dominating wall-time.
I did the top-level parallelization because it was good enough back then, but tidy has grown since.

I'll do a comparison of this branch vs. the file-parallel one on my machine.

jyn514 commented:
oh sorry I forgot to mention, it saved nearly no time at all :/ first image is with this PR and second image is with file-level parallelism.
![image](https://user-images.githubusercontent.com/23638587/210491805-e77253d9-fa21-4c8d-902a-c453ae8f67a3.png)

![image](https://user-images.githubusercontent.com/23638587/210491815-93b38dff-fdd6-4801-aa96-8ce49c913029.png)

the8472 commented:
Yes, it appears so. But it's not saturating my CPU, so it's still 

```
# master
$ ./x test tidy && hyperfine -w 5 '"build/x86_64-unknown-linux-gnu/stage0-tools-bin/rust-tidy" "." "build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "build" "48"'
Benchmark #1: "build/x86_64-unknown-linux-gnu/stage0-tools-bin/rust-tidy" "." "build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "build" "48"
  Time (mean ± σ):     612.7 ms ±  20.5 ms    [User: 2.189 s, System: 1.378 s]
  Range (min … max):   572.2 ms … 643.1 ms    10 runs


# this PR
Benchmark #1: "build/x86_64-unknown-linux-gnu/stage0-tools-bin/rust-tidy" "." "build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "build" "48"
  Time (mean ± σ):     607.8 ms ±  32.2 ms    [User: 2.635 s, System: 1.411 s]
  Range (min … max):   559.6 ms … 658.3 ms    10 runs


# file-parallel branch
Benchmark #1: "build/x86_64-unknown-linux-gnu/stage0-tools-bin/rust-tidy" "." "build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "build" "48"
  Time (mean ± σ):     635.7 ms ±  24.6 ms    [User: 3.046 s, System: 1.639 s]
  Range (min … max):   597.7 ms … 687.2 ms    10 runs
```


Thread timeline of this PR, you can see it's waiting on some cargo calls from tidy, the other checks finish much faster:

![image](https://user-images.githubusercontent.com/1065730/210497776-7386f20f-0cc6-405f-94bf-5eba3395b5a2.png)

Same of the parallel version, more threads for the tidy work but it's waiting for cargo anyway:

![image](https://user-images.githubusercontent.com/1065730/210496999-29968f20-65ec-4564-ac9d-5661b5687771.png)

If those cargo calls could be parallelized and sped up then the file-parallel version might end up faster, especially if the work could be spread more evenly across the threads (ignore maybe lacks workstealing?).

jyn514 commented:
Oh, interesting - I think all the cargo calls are in `src/tools/tidy/src/deps.rs`, which is why they're serialized - I think it shouldn't be too hard to `spawn` them all in parallel and then wait on each :)

I don't want to block this PR on that change though, I think it's already a good improvement even without the file-level parallelism.

the8472 commented on src/tools/tidy/src/walk.rs:
```diff
@@ -36,34 +35,39 @@ pub fn filter_dirs(path: &Path) -> bool {
-        }
-        f(&entry, &contents);
+        let contents = t!(fs::read(entry.path()), entry.path());
+        let contents_str = match String::from_utf8(contents) {
```
The previous code reused a string buffer, this allocates a new string for each file. That shouldn't be necessary for the single-threaded walk since it can be used as an iterator.

jyn514 replied:
Sure thing, done.

bors commented:
:umbrella: The latest upstream changes (presumably #106544) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

jyn514 committed:
Reuse allocations between files

jyn514 committed:
Don't walk the `tests/` directories in checks that always skip it

`WalkBuilder` handles top-level paths differently than `fn walk` used
to: it doesn't run the `skip` function to determine if it should be
skipped, instead assuming the top-level function is always included.

This is a reasonable assumption; adapt our code so it doesn't make
pointless calls to `walk`.

the8472 commented:
Is walkdir still used anywhere else in tidy? Otherwise it could be removed from its Cargo.toml

jyn514 commented:
It's used in `mir_opt_tests`: https://github.com/rust-lang/rust/blob/40ba0e84d53f605ccf01836e9c2d27892728ae81/src/tools/tidy/src/mir_opt_tests.rs#L9

the8472 commented:
Would replacing it there reduce build times? If not r=me.

jyn514 commented:
I don't think so, the only files there are `.rs`, `.mir`, `.diff`, and `.html` files, and all of them are looked at (intentionally).

@bors r=the8472 rollup

bors commented:
:pushpin: Commit 98334f6a2d2030221975c457eb038008c76efdcb has been approved by `the8472`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=the8472 98334f6a2d2030221975c457eb038008c76efdcb -->
<!-- homu: {"type":"Approved","sha":"98334f6a2d2030221975c457eb038008c76efdcb","approver":"the8472","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

klensy commented on src/tools/tidy/src/walk.rs:
```diff
@@ -36,34 +35,42 @@ pub fn filter_dirs(path: &Path) -> bool {
     walk_no_read(path, skip, &mut |entry| {
         contents.clear();
-        if t!(File::open(entry.path()), entry.path()).read_to_string(&mut contents).is_err() {
-            contents.clear();
-        }
-        f(&entry, &contents);
+        let mut file = t!(File::open(entry.path()), entry.path());
+        t!(file.read_to_end(&mut contents), entry.path());
+        let contents_str = match std::str::from_utf8(&contents) {
+            Ok(s) => s,
+            Err(_) => return, // skip this file
+        };
+        f(&entry, &contents_str);
```
Why replacing`read_to_string` with `read_to_end` and `from_utf8`?

jyn514 replied:
The behavior on error is different. Before it would call `f` on an empty file, now it skips calling `f` altogether.
